app/src/main/java/com/jshvarts/todoapp/arch/MviViewModel.kt

```diff
@@ -11,18 +16,28 @@ import androidx.lifecycle.ViewModel
  * ```
  */
 abstract class MviViewModel<A : UiAction, S : UiState, E : UiEffect>(
-  private val savedStateHandle: SavedStateHandle? = null
-) : ViewModel(), ActionDispatcher<A> {
-  abstract val initialState: S
-  abstract val savedStateHandleKey: String?
-  abstract val actionHandler: (A) -> Unit
-
-  private val isStateInBundle: Boolean
-    get() = if (savedStateHandleKey != null && savedStateHandle != null) {
-      initialState != savedStateHandle[savedStateHandleKey!!]
-    } else false
-
-  override fun dispatchAction(action: A) {
-    return if (isStateInBundle) Unit else actionHandler.invoke(action)
-  }
-}
\ No newline at end of file
+    private val savedStateHandle: SavedStateHandle? = null
+) : ViewModel() {
+    protected abstract val initialState: S
+
+    private val _uiEffect = MutableSharedFlow<E>()
+    val uiEffect: SharedFlow<E> = _uiEffect.asSharedFlow()
+
+    abstract val uiState: StateFlow<S>
+
+    protected abstract val savedStateHandleKey: String?
+    protected abstract fun handleAction(action: A)
+
+    private val isStateInBundle: Boolean
+        get() = if (savedStateHandleKey != null && savedStateHandle != null) {
+            initialState != savedStateHandle[savedStateHandleKey!!]
+        } else false
+
+    fun dispatchAction(action: A) {
+        return if (isStateInBundle) Unit else handleAction(action)
+    }
+
+    protected suspend fun dispatchEffect(effect: E) {
```

Comment:
`dispatch` term usually refers to coming from the view. There are multiple was to emit to SharedFlow and I dont think we should limit to the suspending emit. see above too

Reply:
Removed.